src/engine.ml

```diff
@@ -1460,15 +1500,15 @@ let wrap_hmac_control now ts mtu session tls_auth key transport outs =
   in
   (session, transport, List.rev outs)
 
-let wrap_tls_crypt_control now ts mtu session (tls_crypt, wkc) needs_wkc key
+let wrap_tls_crypt_control now ts mtu session tls_crypt wkc_opt needs_wkc key
     transport outs =
   let now_ts = ptime_to_ts_exn now in
   (* If we reply with hmac cookie we must split such that the first control
      packet, the Control_wkc has room for the /cleartext/ wkc, and fix the
      packet length afterwards *)
   let session, transport, maybe_out, outs =
-    match (needs_wkc, outs) with
-    | true, (`Control, data) :: rest ->
+    match (wkc_opt, needs_wkc, outs) with
```

Comment:
We could unify `wkc_opt` and `needs_wkc` so we pass always `None` to `wrap_tls_crypt_control` when (current) `needs_wkc` is false